include/ibamr/BrinkmanPenalizationAdvDiff.h

```diff
@@ -0,0 +1,295 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2019 - 2019 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+#ifndef included_BrinkmanPenalizationAdvDiff
+#define included_BrinkmanPenalizationAdvDiff
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibamr/ibamr_enums.h"
+
+#include "ibtk/ibtk_macros.h"
+#include "ibtk/ibtk_utilities.h"
+
+#include "tbox/Database.h"
+#include "tbox/DescribedClass.h"
+#include "tbox/Pointer.h"
+
+#include <limits>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBAMR
+{
+class AdvDiffHierarchyIntegrator;
+} // namespace IBAMR
+namespace SAMRAI
+{
+namespace pdat
+{
+template <int DIM, class TYPE>
+class CellVariable;
+} // namespace pdat
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief BrinkmanPenalizationAdvDiff is an abstract class that provides an interface
+ * to implement Brinkman penalization body force in the advection-diffusion equation
+ * in order to enforce Dirichlet and Neumann boundary conditions on surfaces of rigid
+ * immersed bodies. A single instance of this class is meant to handle all of the Brinkman
+ * penalization zones for multiple transported quantities with various boundary conditions.
+
+ * TODO: A description of the penalization terms for both Dirichlet and Neumann BCs
+ * For Dirichlet BCs, inhomogeneous and homogeneous BCs are treated in the same way.
+ * For Neumann BCs, homogeneous BCs are fairly straightforward, while inhomogeneous BCs require
+ * some additional input by the user. Add a reference to Sakurai's paper
```

Comment:
Would you please finish this TODO?

Reply:
I have added the test for BPAdvDiff which you can find in tests/adv_diff. I have also investigated the test ```wave_tank/nwt_cylinder```. We now use a more appropriate formula for interface width while computing Heaviside/smooth indicator function and this is implemented in ```BrinkmanPenalizationRigidBodyDynamics``` class which was the reason the test was failing. Now I have updated the wave_tank test output.